goaci.go

```diff
@@ -47,189 +27,398 @@ func (v *StringVector) Set(str string) error {
+	keepTmp   bool
+	project   string
+}
 
-	flag.Var(&execOpts, "exec", "Parameters passed to app, can be used multiple times")
+func getOptions() (*options, error) {
+	opts := &options{}
+
+	// --go-binary
+	goDefaultBinaryDesc := "Go binary to use"
+	gocmd, err := exec.LookPath("go")
 	if err != nil {
-		goDefaultBinaryDesc = "Go binary to use (default: none found in $PATH, so it must be provided)"
+		goDefaultBinaryDesc += " (default: none found in $PATH, so it must be provided)"
 	} else {
-		goDefaultBinaryDesc = "Go binary to use (default: whatever go in $PATH)"
+		goDefaultBinaryDesc += " (default: whatever go in $PATH)"
 	}
-	flag.StringVar(&goBinaryOpt, "go-binary", gocmd, goDefaultBinaryDesc)
-	flag.StringVar(&goPathOpt, "go-path", "", "Custom GOPATH (default: a temporary directory)")
+	flag.StringVar(&opts.goBinary, "go-binary", gocmd, goDefaultBinaryDesc)
+
+	// --exec
+	flag.Var(&opts.exec, "exec", "Parameters passed to app, can be used multiple times")
+
+	// --go-path
+	flag.StringVar(&opts.goPath, "go-path", "", "Custom GOPATH (default: a temporary directory)")
+
+	// --use-binary
+	flag.StringVar(&opts.useBinary, "use-binary", "", "Which executable to put in ACI image")
+
+	// --asset
+	flag.Var(&opts.assets, "asset", "Additional assets, can be used multiple times; format: <path in ACI>"+ListSeparator()+"<local path>; placeholders like <GOPATH> and <PROJPATH> can be used there as well")
+
+	// --keep-tmp
+	flag.BoolVar(&opts.keepTmp, "keep-tmp", false, "Do not delete temporary directory used for creating ACI")
 	flag.Parse()
+
+	args := flag.Args()
+	if len(args) != 1 {
+		return nil, fmt.Errorf("Expected exactly one project to build, got %d", len(args))
+	}
+	opts.project = args[0]
+	if opts.goBinary == "" {
+		return nil, fmt.Errorf("Go binary not found")
+	}
+
+	return opts, nil
+}
+
+type pathsAndNames struct {
+	tmpDirPath  string
+	goPath      string
+	goRootPath  string
+	projectPath string
+	fakeGoPath  string
+	goBinPath   string
+	aciDirPath  string
+	rootFSPath  string
+	goExecPath  string
+
+	imageFileName string
+	imageACName   string
+}
+
+// getGoPath returns go path and fake go path. The former is either in
+// /tmp (which is a default) or some other path as specified by
+// --go-path parameter. The latter is always in /tmp.
+func getGoPath(opts *options, tmpDir string) (string, string) {
+	fakeGoPath := filepath.Join(tmpDir, "gopath")
+	if opts.goPath == "" {
+		return fakeGoPath, fakeGoPath
+	}
+	return opts.goPath, fakeGoPath
+}
+
+// getNamesFromProject returns project name, image ACName and ACI
+// filename. Names depend on whether project has several binaries. For
+// project with single binary (github.com/appc/goaci) returned values
+// would be: github.com/appc/goaci, github.com/appc/goaci and
+// goaci.aci. For project with multiple binaries
+// (github.com/appc/spec/...) returned values would be (assuming ace
+// as selected binary): github.com/appc/spec, github.com/appc/spec-ace
+// and spec-ace.aci.
```

Comment:
My impression is that ACI image is basically a package of one standalone application and its assets, in this case - etcd daemon. If you want more binaries, create more ACI images. That's why I'm appending binary name to project name if it produces more than one binary. The case would be even worse if you have project `foo` that builds two daemons `bar` and `baz`. I suppose it would be better to create differently named images for daemon `bar` (like `foo-bar.aci`) and for daemon `baz` (like `foo-baz.aci`) instead of having `foo.aci` and wondering if it runs `bar` or `baz` daemon.

If a binary forks and executes other binaries then it is builder's responsibility to add them as an asset.

As of `./build` - sounds like user has to run `goaci github.com/coreos/etcd/...` to get both `etcd` and `etcdctl` binary. I haven't tested that.